eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/Imp/UtxoSpec.hs

```diff
@@ -0,0 +1,140 @@
+      -- compute the difference between the current-era minFee and that computed in pre-Conway eras
+      minFeeDiff <- conwayDiffMinFee tx
+      -- check that the difference is the sum of the sizes of the passed reference scripts
+      minFeeDiff
+        `shouldBe` Foldable.foldMap (\s -> scriptSize s <×> refScriptFee) refScriptInToScripts
+
+    conwayDiffMinFee :: Tx era -> ImpTestM era Coin
+    conwayDiffMinFee tx = do
+      utxo <- getUTxO
+      pp <- getsNES $ nesEsL . curPParamsEpochStateL
+      pure $ getMinFeeTxUtxo pp tx utxo <-> getShelleyMinFeeTxUtxo pp tx
+
+    createScriptUtxo :: NativeScript era -> ImpTestM era (TxIn (EraCrypto era))
+    createScriptUtxo script = do
+      scriptAddr <- addScriptAddr script
+      tx <-
+        submitTx . mkBasicTx $
+          mkBasicTxBody
+            & outputsTxBodyL @era
+              .~ SSeq.fromList [mkBasicTxOut @era scriptAddr (inject (Coin 1000))]
+      pure $ txInAt 1 tx
+
+    createRefScriptsUtxos :: [NativeScript era] -> ImpTestM era (Map.Map (TxIn (EraCrypto era)) (NativeScript era))
+    createRefScriptsUtxos scripts = do
+      rootOut <- snd <$> lookupImpRootTxOut
+      let outs =
+            scripts
+              <&> ( \s ->
+                      mkBasicTxOut @era (rootOut ^. addrTxOutL) (inject (Coin 100))
+                        & referenceScriptTxOutL @era .~ SJust (fromNativeScript s)
+                  )
+      tx <-
+        submitTx . mkBasicTx $
+          mkBasicTxBody
+            & outputsTxBodyL @era
+              .~ SSeq.fromList outs
+      let refIns = (\i -> txInAt (fromIntegral i) tx) <$> [1 .. length scripts]
+      pure $ Map.fromList $ refIns `zip` scripts
+
+    spendScriptUsingRefScripts :: TxIn (EraCrypto era) -> [TxIn (EraCrypto era)] -> ImpTestM era (Tx era)
+    spendScriptUsingRefScripts scriptIn refIns = do
+      submitTx . mkBasicTx $
+        mkBasicTxBody
+          & inputsTxBodyL @era .~ Set.singleton scriptIn
+          & referenceInputsTxBodyL @era .~ Set.fromList refIns
+
+    mkScript :: ImpTestM era (NativeScript era)
+    mkScript = do
+      requiredKeyHash <- freshKeyHash
+      pure $ RequireAllOf (SSeq.singleton (RequireSignature @era requiredKeyHash))
+
+    addScriptAddr :: NativeScript era -> ImpTestM era (Addr (EraCrypto era))
+    addScriptAddr script = do
+      kpStaking1 <- lookupKeyPair =<< freshKeyHash
+      scriptHash <- impAddNativeScript script
+      pure $ mkScriptAddr scriptHash kpStaking1
+
+    scriptSize :: NativeScript era -> Int
+    scriptSize = SBS.length . getMemoRawBytes
+
+    setRefScriptFee :: ImpTestM era Coin
+    setRefScriptFee = do
+      let refScriptFee = Coin 10
+      modifyPParams $ ppMinFeeRefScriptCoinsPerByteL .~ CoinPerByte refScriptFee
+      pure refScriptFee
+
+    txInAt :: Integer -> Tx era -> TxIn (EraCrypto era)
+    txInAt index tx =
+      let txid = txIdTx tx
+       in mkTxInPartial txid index
```

Comment:
It could be a useful function for testing in general, so maybe we should put it into core testlib. Also type signature could be more general and with `HasCallStack`
```suggestion
    txInAt :: (HasCallStack, Integral i) => i -> Tx era -> TxIn (EraCrypto era)
    txInAt index tx =
      let txid = txIdTx tx
       in mkTxInPartial txid (toInteger index)
```

Reply:
Sounds like as good of a place as any, at least for the time being. Maybe we'll find a better home for it in the future.
Thank you.